### PR TITLE
Implementation of GoDaddy provider

### PR DESCRIPTION
**Pull request by adferrand**
To continue my journey on Lexicon, I will focus today on another widly-used DNS provider : GoDaddy.

This PR includes the GoDaddy provider to make all CRUD operations on a DNS zone handled by GoDaddy.

Test integrations have been constructed, and executed with success. It uses the OTE GoDaddy server, which allows tests without a paid account.

I am available for further discussions about merging this PR.

Regards,
Adrien Ferrand

### CODE REVIEW

**Comment by CapsuleCD** [comment]
Hi.

I'm an automated pull request bot named [CapsuleCD](https://www.capsulecd.com). I handle testing, versioning and package releases for this project. 
- If you're the owner of this repo, you can click the button below to kick off the CapsuleCD build pipeline and create an automated release.'
- If not, don't worry, someone will be by shortly to check on your pull request. 

[CapsuleCD](https://www.capsulecd.com/project/github/AnalogJ/lexicon/pullrequests/131)

---
If you're interested in learning more about [CapsuleCD](http://www.github.com/AnalogJ/capsulecd), or adding it to your project, you can check it out [here](https://www.capsulecd.com)

**Comment by AnalogJ** [comment]
Awesome work! I remember looking around for a Sandbox environment for GoDaddy, but never saw anything like the OTE environment you found. Thanks for pointing it out :)

I'll take a look at your PR later today and get back to you.

File: lexicon/providers/godaddy.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    def update_record(self, identifier, type=None, name=None, content=None):
+        # With GoDaddy API, creating a record for given type and name is the same
+        #   than updating the record.
+        return self.create_record(type, name, content)
+
+    def delete_record(self, identifier=None, type=None, name=None, content=None):
+        domain = self.options.get('domain')
+
+        if not type:
+            raise Exception('ERROR: type must be setted')
```
Comment (AnalogJ) [inline lexicon/providers/godaddy.py]: these exceptions should probably be:

`raise Exception('Error: type is required')` just to be consistent. 

I'll probably have to think more about a consistent Exception pattern and input validation.

File: lexicon/providers/godaddy.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        # However it provides a way to get and update all records of a zone.
+        # So :
+        #  - we get all the records,
+        #  - we filter the array to remove the record to be deleted,
+        #  - then we push back the filtered array to set the zone without the record to be deleted.
+        # And yes, we could limit the operation to a given type record (eg. TXT, there is an URL
+        #   for that), but GoDaddy refuses to push back an empty set of a given type (yep, you
+        #   cannot remove all your TXT with this URL, ultra weird).
+        # It is likely to happen during a DNS challenge, as all TXT should be removed at the end.
+        # So operating on all the zone avoid empty sets (there will always at least NS entries).
```
Comment (AnalogJ) [inline lexicon/providers/godaddy.py]: wow, thats an awful api. Thanks for figuring this all out :)

File: tests/fixtures/cassettes/godaddy/IntegrationTests/test_Provider_authenticate.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      Accept: [application/json]
+      Accept-Encoding: ['gzip, deflate']
+      Connection: [keep-alive]
+      Content-Length: ['2']
+      Content-Type: [application/json]
+      User-Agent: [python-requests/2.11.1]
+    method: GET
+    uri: https://api.ote-godaddy.com/v1/domains/000.biz
+  response:
+    body: {string: !!python/unicode '{"domainId":1002111,"domain":"000.biz","status":"TRANSFERRED_OUT","expires":"2016-06-14T23:59:59Z","expirationProtected":false,"holdRegistrar":false,"locked":true,"privacy":false,"renewAuto":true,"renewable":false,"transferProtected":false,"createdAt":"2015-06-15T13:10:43Z","authCode":"C36C3C6M26XE4DB5","subaccountId":"857256"}'}
```
Comment (AnalogJ) [inline tests/fixtures/cassettes/godaddy/IntegrationTests/test_Provider_authenticate.yaml]: This "authCode" key in the body looks odd. Are you sure its not a variable you need to filter?

File: lexicon/providers/godaddy.py (hunk 2 above)
Comment (adferrand) [inline lexicon/providers/godaddy.py]: Yep really weird. Even more when you consider their approach about "merging" insert and update functionalities which is in fact pretty relevant.

File: tests/fixtures/cassettes/godaddy/IntegrationTests/test_Provider_authenticate.yaml (hunk 3 above)
Comment (adferrand) [inline tests/fixtures/cassettes/godaddy/IntegrationTests/test_Provider_authenticate.yaml]: GoDaddy provides in plain text and hard coded a key and secret to use the API on OTE servers. Plenty of domains are registered on this account, and they are clearly the result of tests. I think that any data exposed here is non critical.

File: lexicon/providers/godaddy.py (hunk 1 above)
Comment (adferrand) [inline lexicon/providers/godaddy.py]: I corrected the text for theses exceptions.

Comment (AnalogJ) [inline lexicon/providers/godaddy.py]: thanks

File: tests/fixtures/cassettes/godaddy/IntegrationTests/test_Provider_authenticate.yaml (hunk 3 above)
Comment (AnalogJ) [inline tests/fixtures/cassettes/godaddy/IntegrationTests/test_Provider_authenticate.yaml]: Fair enough.

**Comment by AnalogJ** [comment]
LGTM. Merging.

**Pull request merged**